plugin/core/main.py

```diff
@@ -6,24 +6,42 @@
 from .logging import set_debug_logging, set_exception_logging
 from .panels import destroy_output_panels, ensure_panel, PanelName
 from .popups import popups
-from .registry import windows, load_handlers, unload_sessions
+from .registry import windows, unload_sessions
+from .sessions import Session
 from .settings import settings, load_settings, unload_settings
-from .typing import Optional
+from .typing import List, Optional, Type
+from .windows import register_session_type
 
 
 def ensure_server_panel(window: sublime.Window) -> Optional[sublime.View]:
     return ensure_panel(window, PanelName.LanguageServers, "", "", "Packages/LSP/Syntaxes/ServerLog.sublime-syntax")
 
 
+def get_final_subclasses(derived: List[Type], results: List[Type]) -> None:
+    for d in derived:
+        d_subclasses = d.__subclasses__()
+        if len(d_subclasses) > 0:
+            get_final_subclasses(d_subclasses, results)
+        else:
+            results.append(d)
+
+
 def startup() -> None:
     load_settings()
     popups.load_css()
     set_debug_logging(settings.log_debug)
     set_exception_logging(True)
+    final_subclasses = []  # type: List[Type[Session]]
+    get_final_subclasses(Session.__subclasses__(), final_subclasses)
+    for session_type in final_subclasses:
+        # Unfortunately we need to do this ourselves instead of helper packages registering themselves in plugin_loaded
+        # and unregistering themselves in plugin_unloaded. This is because this base LSP package is loaded before
+        # the helper packages. If the helper packages were loaded before this package, we would not have to do this.
+        # Perhaps we can come up with a scheme to load the helper packages before this package?
+        register_session_type(session_type)
```

Comment:
I'm not sure I understand the reasoning?

When saying one package is loaded before another, you mean the `plugin_loaded` handler, not the importing of the python files itself?

Why would package not be able to register itself regardless if it's called before or after LSP?
It would import some global `register` function from LSP and call it. If LSP hasn't loaded yet, it would be put in the queue and registered later.

Reply:
I admit my comments don't explain the problem.

The problem is that at startup we also check which sessions to start for the open views. So we fetch relevant session configs and then start some sessions.

The problem is that when that happens, LSP-* are too late for registering their custom session.

Reply:
So the problem is related to current architecture but should be fixable right?
Another matter whether it's fixable within this PR.

Reply:
I'm keeping the behavior the same for now. I'm not sure what the solution should be.